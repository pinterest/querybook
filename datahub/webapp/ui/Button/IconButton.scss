.IconButton {
    cursor: pointer;
    padding: 10px;
    color: var(--icon-color);
    transition: color 0.2s ease-out;
    transition: fill 0.2s ease-out;
    display: inline-flex;
    position: relative;

    &.invert-circle {
        background-color: var(--color-primary-3);
        padding: 4px;
        border-radius: 50%;
    }

    &.no-padding {
        padding: 0px;
    }
    &:hover {
        color: var(--icon-hover-color);
    }
    &.active {
        color: var(--icon-active-color);
    }
    &.fill {
        svg {
            fill: var(--icon-color);
        }
        &:hover {
            svg {
                fill: var(--icon-hover-color);
            }
        }
        &.active {
            svg {
                fill: var(--icon-active-color);
            }
        }
    }

    &.disabled {
        cursor: not-allowed;
    }

    &.with-title {
        padding-bottom: 0px;
    }

    .ping,
    .ping-message {
        position: absolute;
        cursor: default;
        border-radius: 100px;
        background-color: var(--color-accent-bg);
    }

    .ping {
        right: 8px;
        top: 8px;

        width: 12px;
        height: 12px;
    }

    .ping-message {
        right: 4px;
        top: 4px;
        padding: 1px 6px 0px 6px;

        color: var(--color-accent-text);
        font-size: var(--xsmall-text-size);
        font-weight: var(--bold-font);
    }

    .icon-title {
        font-size: var(--xxxsmall-text-size);
    }
}
