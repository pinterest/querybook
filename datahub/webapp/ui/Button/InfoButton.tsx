import React from 'react';
import { PopoverLayout, Popover, PopoverDirection } from 'ui/Popover/Popover';
import './InfoButton.scss';

export interface IInfoButtonProps {
    layout?: PopoverLayout;
    popoverClassName?: string;
}

export const InfoButton: React.FunctionComponent<IInfoButtonProps> = ({
    layout = ['top'] as [PopoverDirection],

    children,
    popoverClassName,
}) => {
    const [showInfo, setShowInfo] = React.useState(false);
    const buttonRef = React.useRef<HTMLSpanElement>();

    const popover = showInfo && (
        <Popover
            anchor={buttonRef.current}
            layout={layout}
            onHide={() => setShowInfo(false)}
            resizeOnChange
        >
            <div
                className={
                    popoverClassName
                        ? `info-button-popover ${popoverClassName}`
                        : 'info-button-popover '
                }
            >
                {children}
            </div>
        </Popover>
    );

    return (
        <span className="InfoButton" ref={buttonRef}>
            <i
                className="fas fa-info-circle"
                onClick={() => setShowInfo(!showInfo)}
            />
            {popover}
        </span>
    );
};
