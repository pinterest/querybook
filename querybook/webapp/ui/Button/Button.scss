@import './../../scss_variables.scss';

.Button {
    display: inline-flex;
    margin: 0px 4px;
    border: var(--border);
    border-radius: var(--border-radius);
    text-align: center;
    cursor: pointer;
    padding: 4px 8px;
    background-color: inherit;
    align-items: center;
    white-space: nowrap;
    position: relative;

    .Icon {
        margin-left: -4px;
        svg {
            padding: 16%;
        }
    }

    .Icon + span {
        margin-left: 4px;
    }

    &.icon-only {
        .Icon {
            margin-left: 0px;
        }
    }

    // special buttons
    &.inline-text {
        @include pushable();
        color: inherit;
        font-size: inherit;
        font-weight: 700;
        text-transform: uppercase;

        .Icon {
            svg {
                width: 20px;
                height: 20px;
            }
        }

        &:hover {
            background-color: var(--hover-bg-color);
        }
    }

    &.soft {
        user-select: none;
        background: var(--light-bg-color);
        color: var(--light-text-color);
        font-weight: bold;
        border: none;

        &:hover {
            color: var(--text-color);
        }
    }

    &.full-width {
        display: flex;
        justify-content: center;
        margin: 0px;
        border: none;
        border-radius: 0px;
        padding: 8px;
        background-color: var(--light-bg-color);
        font-weight: bold;
        text-transform: uppercase;
    }

    &.cancel {
        border-color: var(--color-false-dark);
        color: var(--color-false-dark);
        font-weight: 500;
    }

    &.confirm {
        border-color: var(--color-true-dark);
        color: var(--color-true-dark);
        font-weight: 500;
    }

    // props
    &.small {
        margin: 0px;
        padding: 1px 4px;
    }

    &.disabled {
        cursor: not-allowed;
        background-color: var(--color-null);
        color: var(--light-text-color);
        border-color: var(--color-null);
    }

    &.borderless {
        border: none;

        &:hover {
            background-color: var(--hover-bg-color);
        }
    }
    &.inverted {
        background-color: var(--color-accent-bg);
        color: var(--color-accent-text);
        font-weight: 500;
    }

    &.attached-right {
        border-top-right-radius: 0px;
        border-bottom-right-radius: 0px;
        margin-right: 0px;
        + .Button {
            border-left: 0px;
        }
    }
    &.attached-left {
        border-top-left-radius: 0px;
        border-bottom-left-radius: 0px;
        margin-left: 0px;
    }

    &.pushable {
        @include pushable();
    }

    .ping-message {
        position: absolute;
        cursor: default;
        border-radius: 100px;
        background-color: var(--color-accent-bg);

        right: -8px;
        top: -8px;

        padding: 1px 6px 0px 6px;

        color: var(--color-accent-text);
        font-size: var(--xsmall-text-size);
        font-weight: var(--bold-font);
    }
}
