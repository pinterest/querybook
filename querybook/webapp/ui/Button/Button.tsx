import React from 'react';
import clsx from 'clsx';

import { Icon, IIconProps } from 'ui/Icon/Icon';
import { StyledButton, ISharedButtonProps } from './StyledButton';
import { withBoundProps } from 'lib/utils/react-bind';
import {
    ButtonColorType,
    ButtonThemeType,
    computeStyleButtonProps,
} from './ButtonTheme';

export type ButtonProps = React.HTMLAttributes<HTMLSpanElement> &
    ISharedButtonProps & {
        icon?: string | React.ComponentType<IIconProps>;
        title?: string;
        className?: string;

        color?: ButtonColorType;
        theme?: ButtonThemeType;

        disabled?: boolean;
        isLoading?: boolean;
        active?: boolean;
        ping?: string;
    };

const defaultProps: ButtonProps = {
    className: '',
    color: 'default',
    theme: 'outline',
    size: 'medium',
};

export const Button = React.forwardRef<HTMLSpanElement, ButtonProps>(
    (props, ref) => {
        const {
            children,
            icon,
            title,
            className,
            color,
            theme,
            disabled,
            onClick,
            active = false,
            isLoading = false,
            ping = null,
            ...elementProps
        } = props;

        const iconDOM = isLoading ? (
            <Icon name="loader" />
        ) : (
            icon && (typeof icon === 'string' ? <Icon name={icon} /> : icon)
        );
        const textDOM = title && <span>{title}</span>;

        const buttonOnClick = disabled ? null : onClick;
        const themeProps = computeStyleButtonProps(color, theme);

        const buttonClassName = clsx({
            Button: true,
            [className]: !!className,
            active,
            'flex-row': Boolean(iconDOM && textDOM),
            'icon-only': Boolean(iconDOM && !textDOM),
        });

        const pingDOM = !ping ? null : (
            <div className="ping-message">{ping}</div>
        );
        return (
            <StyledButton
                className={buttonClassName}
                disabled={disabled}
                {...themeProps}
                {...elementProps}
                ref={ref}
                onClick={buttonOnClick}
            >
                {pingDOM}
                {iconDOM}
                {textDOM}
                {children}
            </StyledButton>
        );
    }
);

Button.defaultProps = defaultProps;
Button.displayName = 'Button';

export const TextButton = withBoundProps(Button, {
    theme: 'text',
    fontWeight: '700',
    uppercase: true,
    pushable: true,
});

export const SoftButton = withBoundProps(Button, {
    theme: 'fill',
    color: 'light',
    fontWeight: '700',
    pushable: true,
});

export type ButtonType = 'soft' | 'text' | 'default';

const buttonComponentByType: Record<
    ButtonType,
    React.ComponentType<ButtonProps>
> = {
    soft: SoftButton,
    text: TextButton,
    default: Button,
};
export function getButtonComponentByType(buttonType?: ButtonType) {
    buttonType = buttonType ?? 'default';
    return buttonComponentByType[buttonType];
}
