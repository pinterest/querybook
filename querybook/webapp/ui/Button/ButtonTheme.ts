import { IStyledButtonThemeProps } from './StyledButton';

export const ButtonColors = [
    'confirm',
    'cancel',
    'accent',
    'light',
    'default',
] as const;
export const ButtonThemes = ['outline', 'text', 'fill'] as const;

export type ButtonColorType = typeof ButtonColors[number];
export type ButtonThemeType = typeof ButtonThemes[number];

interface IButtonColorConfig {
    primary: string; // text and icon
    primaryHover?: string;
    secondary: string; // Background and border
    secondaryHover?: string;
}

const buttonThemeToProps: Record<ButtonColorType, IButtonColorConfig> = {
    confirm: {
        primary: 'var(--color-true-dark)',
        secondary: 'var(--color-true-dark)',
        secondaryHover: 'var(--color-true)',
    },
    cancel: {
        primary: 'var(--color-false-dark)',
        secondary: 'var(--color-false-dark)',
        secondaryHover: 'var(--color-false)',
    },
    accent: {
        primary: 'var(--color-accent-text)',
        secondary: 'var(--color-accent-bg)',
    },
    light: {
        primary: 'var(--light-text-color)',
        primaryHover: 'var(--text-color)',
        secondary: 'var(--light-bg-color)',
        secondaryHover: 'var(--hover-bg-color)',
    },
    default: {
        primary: 'var(--text-color)',
        primaryHover: 'var(--dark-text-color)',
        secondary: 'var(--border-color)',
        secondaryHover: 'var(--hover-border-color)',
    },
};

export function computeStyleButtonProps(
    colorType: ButtonColorType,
    theme: ButtonThemeType
) {
    const colorConfig = buttonThemeToProps[colorType];
    const secondaryIsPrimary = colorConfig.secondary === colorConfig.primary;
    const themeProps: IStyledButtonThemeProps = {};

    if (theme === 'fill') {
        if (!secondaryIsPrimary) {
            themeProps.color = colorConfig.primary;
            themeProps.hoverColor =
                colorConfig.primaryHover || colorConfig.primary;
            themeProps.bgColor = colorConfig.secondary;
            themeProps.hoverBgColor =
                colorConfig.secondaryHover || colorConfig.secondary;
        } else {
            themeProps.color = 'var(--bg-color)';
            themeProps.hoverColor = 'var(--light-bg-color)';
            themeProps.bgColor = colorConfig.primary;
            themeProps.hoverBgColor =
                colorConfig.primaryHover || colorConfig.primary;
        }
    } else if (theme === 'outline') {
        themeProps.color = colorConfig.primary;
        themeProps.hoverColor = colorConfig.primaryHover || colorConfig.primary;

        themeProps.bgColor = 'var(--bg-color)';
        themeProps.borderColor = colorConfig.secondary;
        themeProps.hoverBorderColor =
            colorConfig.secondaryHover || colorConfig.secondary;
    } else if (theme === 'text') {
        themeProps.color = colorConfig.primary;
        themeProps.hoverColor = colorConfig.primaryHover || colorConfig.primary;
        themeProps.bgColor = 'transparent';
        themeProps.hoverBgColor = 'var(--light-bg-color)';
    }

    return themeProps;
}
