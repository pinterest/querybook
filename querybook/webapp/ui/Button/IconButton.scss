.IconButton {
    cursor: pointer;
    padding: 10px;
    color: var(--icon);
    transition: color 0.2s ease-out;
    transition: fill 0.2s ease-out;
    display: inline-flex;
    position: relative;

    &.invert-circle {
        background-color: var(--bg-lightest);
        padding: 4px;
        border-radius: 50%;

        &:hover {
            background-color: var(--bg-hover);
        }
    }

    &.no-padding {
        padding: 0px;
    }

    &:not(.disabled) {
        &:hover {
            color: var(--icon-hover);
        }
        &.active {
            color: var(--color-accent-dark);
        }
    }

    &.disabled {
        cursor: not-allowed;
    }

    &.with-title {
        width: 100%;
        padding: 10px 4px 8px;
        display: flex;
        justify-content: center;
    }

    .ping,
    .ping-message {
        position: absolute;
        cursor: default;
        border-radius: 100px;
        background-color: var(--color-accent-dark);
    }

    .ping {
        right: 8px;
        top: 8px;

        width: 12px;
        height: 12px;
    }

    .ping-message {
        right: 4px;
        top: 4px;
        padding: 1px 6px 0px 6px;
        color: var(--color-accent-lightest-0);
        font-weight: var(--bold-font);
    }

    .icon-title {
        padding-top: 2px;
    }
}
