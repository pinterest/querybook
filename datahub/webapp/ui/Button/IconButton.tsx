import React from 'react';
import classNames from 'classnames';

import { TooltipDirection } from 'const/tooltip';

import { Icon } from 'ui/Icon/Icon';

import './IconButton.scss';

// A simple Bulma-based button
export interface IconButtonProps {
    icon: string;
    className?: string;
    onClick?: (event?: React.MouseEvent<HTMLSpanElement, MouseEvent>) => any;

    tooltip?: React.ReactNode;
    tooltipPos?: TooltipDirection;

    disabled?: boolean;
    active?: boolean;
    noPadding?: boolean;
    fill?: boolean;
    size?: string | number;
    ping?: boolean | string;
    // Bug: somehow typescript can't auto detect this field
    // after forwardRef
    children?: React.ReactNode;
}

export const IconButton = React.forwardRef<HTMLAnchorElement, IconButtonProps>(
    (
        {
            icon,
            children,
            onClick,
            className = '',
            tooltip,
            tooltipPos = 'up',
            disabled,

            active,
            size,
            noPadding,
            fill,
            ping,
        },
        ref
    ) => {
        const iconButtonProps = {
            ref,
            onClick: disabled ? null : onClick,
        };

        if (tooltip) {
            iconButtonProps['aria-label'] = tooltip;
            iconButtonProps['data-balloon-pos'] = tooltipPos;
        }

        const iconButtonClassname = classNames({
            IconButton: true,
            active,
            'no-padding': noPadding,
            fill,
            disabled,
            [className]: className,
        });

        return (
            <span {...iconButtonProps} className={iconButtonClassname}>
                {!ping ? null : typeof ping === 'boolean' ? (
                    <div className="ping" />
                ) : (
                    <div className="ping-message">{ping}</div>
                )}
                <Icon name={icon} size={size} fill={fill} /> {children}
            </span>
        );
    }
);
