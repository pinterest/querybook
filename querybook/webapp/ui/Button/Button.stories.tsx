import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { Button, ButtonProps, SoftButton, TextButton } from './Button';
import { ButtonColors, ButtonThemes } from './ButtonTheme';

export default {
    title: 'Button/Button',
    decorators: [centered],
};

export const _Button = ({ ...args }: ButtonProps) => <Button {...args} />;
_Button.args = {
    icon: 'zap',
    title: 'Button',
    color: 'default',
    theme: 'outline',
    disabled: false,
    isLoading: false,

    pushable: false,
    size: 'medium',
    fullWidth: false,
    uppercase: false,
    attached: '',
    fontWeight: '',
    ping: '',
};

_Button.argTypes = {
    color: {
        control: {
            type: 'select',
            options: ButtonColors,
        },
    },
    theme: {
        control: {
            type: 'select',
            options: ButtonThemes,
        },
    },
    size: {
        control: {
            type: 'select',
            options: ['small', 'medium'],
        },
    },
    attached: {
        control: {
            type: 'select',
            options: ['', 'left', 'right'],
        },
    },
};

export const ButtonThemeColorExamples = () => (
    <>
        {ButtonThemes.map((theme) => (
            <div key={theme} className="mv12">
                {ButtonColors.map((color) => (
                    <Button key={theme + color} theme={theme} color={color}>
                        {theme} {color}
                    </Button>
                ))}
            </div>
        ))}
    </>
);

export const ButtonStylesExamples = () => (
    <>
        <div className="mb12">
            <Button pushable>Pushable Button</Button>
        </div>

        <div className="mb12">
            <Button disabled>Disabled Button</Button>
        </div>

        <div className="mb12">
            <Button fullWidth>Full Width Button</Button>
        </div>

        <div className="mb12">
            <Button uppercase>Uppercase Button</Button>
        </div>

        <div className="mb12">
            <Button attached="right">Button Attached Right</Button>
            <Button attached="left">Button Attached Left</Button>
        </div>

        <div className="mb12">
            <Button size="small">Small Button</Button>
        </div>
    </>
);
