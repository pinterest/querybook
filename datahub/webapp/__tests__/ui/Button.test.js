import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { Button } from '../../ui/Button/Button';

it('renders without crashing', () => {
    shallow(<Button title="test" onClick={() => null} />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        let wrapper = shallow(<Button title="test" onClick={() => null} />);
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - disbled', () => {
        let wrapper = shallow(
            <Button title="test" onClick={() => null} disabled />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - typed', () => {
        let wrapper = shallow(
            <Button title="test" onClick={() => null} type="soft" />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - borderless', () => {
        let wrapper = shallow(
            <Button title="test" onClick={() => null} borderless />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - attachedRight', () => {
        let wrapper = shallow(
            <Button title="test" onClick={() => null} attachedRight />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - isLoading', () => {
        let wrapper = shallow(
            <Button title="test" onClick={() => null} isLoading />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot - mixed', () => {
        let wrapper = shallow(
            <Button
                title="test"
                onClick={() => null}
                pushable
                transparent
                small
                inverted
                attachedLeft
            />
        );
        let serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(
            <Button title="test" onClick={() => null} />
        );
        expect(output).toMatchSnapshot();
    });
});
