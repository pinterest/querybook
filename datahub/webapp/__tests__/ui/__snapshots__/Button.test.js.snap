// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot - attachedRight 1`] = `
<span
  className="Button  attached-right"
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot - borderless 1`] = `
<span
  className="Button  borderless"
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot - disbled 1`] = `
<span
  className="Button  disabled"
  onClick={null}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot - isLoading 1`] = `
<span
  className="Button  flex-row"
  onClick={[Function]}
>
  <Memo()
    name="loader"
  />
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot - mixed 1`] = `
<span
  className="Button  attached-left transparent pushable small inverted"
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot - typed 1`] = `
<span
  className="Button soft"
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches enzyme snapshots matches snapshot 1`] = `
<span
  className="Button "
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<span
  className="Button "
  onClick={[Function]}
>
  <span>
    test
  </span>
</span>
`;
